impl/client/src/main/javascript/web/pentaho/visual/base/modelAdapter.js

```diff
@@ -589,77 +604,102 @@ define([
-          operands = [];
-
-          filter.operands.each(function(operandFilter) {
-            operandFilter = __transformFilter.call(this, operandFilter, internalData, toExternal);
-            if(operandFilter !== null) {
-              operands.push(operandFilter);
-            }
-          }.bind(this));
-
-          return new OrFilter({operands: operands});
-        }
+        switch(filter.kind) {
+          case "true":
+          case "false":
+          case "or":
+            return null;
+          case "and":
+            // Collect and replace all isEqual children.
+            equalsMap = null;
+            var operands = [];
+
+            var filterOperands = filter.operands;
+            var filterOperandsCount = filterOperands.count;
+            for(var iOperand = 0; iOperand < filterOperandsCount; iOperand++) {
+              var operandFilter = filterOperands.at(iOperand);
+
+              if(operandFilter.kind === "isEqual") {
+                if(equalsMap === null) {
+                  equalsMap = Object.create(null);
+                }
 
-        if(filter.kind === "and") {
-          // Collect and replace all isEqual children.
-          equalsMap = null;
-          operands = [];
+                // The same property cannot have different values at the same time therefore return False
+                if(equalsMap[operandFilter.property] !== undefined &&
+                  equalsMap[operandFilter.property] !== operandFilter.value) {
+                  return FalseFilter.instance;
+                }
 
-          filter.operands.each(function(operandFilter) {
-            if(operandFilter.kind === "isEqual") {
-              if(equalsMap === null) {
-                equalsMap = Object.create(null);
+                equalsMap[operandFilter.property] = operandFilter.value;
+              } else {
+                // visit other filter types
+                operands.push(operandFilter.visit(__transformFilter.bind(this)));
               }
+            }
 
-              equalsMap[operandFilter.property] = operandFilter.value;
-            } else {
-              // Pass-through other filter kinds.
-              operands.push(operandFilter);
+            // if isEqual operands found for And filter
+            if(equalsMap !== null) {
+              // Map internal/external values to external/internal values.
+              equalsMap = this.modelAdapter.__convertValuesMap(equalsMap, this.toExternal);
+
+              operands.push.apply(operands, Object.keys(equalsMap).map(function(propName) {
+                return dataUtil.createFilterIsEqualFromCell(
+                  this.internalData,
+                  propName,
+                  equalsMap[propName],
+                  __context
+                );
+              }, this));
             }
-          });
 
-          if(equalsMap === null) {
-            return filter;
-          }
+            return new AndFilter({operands: operands});
```

Comment:
What happens if `equalsMap` is empty? Is an empty `and` equivalent to `true`, like in `isEqual` case bellow?